enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,491 @@
+---
+title: set-replace-append-delete-deny-headers
+authors:
+  - "@miheer"
+reviewers:
+  - "@frobware"
+  - "@gcs278"
+  - "@Miciah"
+approvers:
+  - "@frobware"
+  - "@gcs278"
+  - "@Miciah"
+api-approvers:
+  - "@JoelSpeed"
+  - "@deads2k"
+  - "@frobware"
+  - "@gcs278"
+  - "@Miciah"
+creation-date: 2022-12-02
+last-updated: 2022-12-14
+tracking-link:
+  - "https://issues.redhat.com/browse/NE-982"
+  - "https://issues.redhat.com/browse/RFE-464"
+status: TODO
+see-also:
+  - ""
+replaces:
+superseded-by:
+---
+
+# Set, Delete, Deny HTTP Response headers
+
+## Summary
+
+This enhancement extends the IngressController API and Route API to allow the user to set or delete
+HTTP response headers. Using the new API, users may specify a list of
+HTTP response headers either via IngressController API or Route API that should be set or deleted.
+Also, this enhancement creates an immutable(hard-coded) list of HTTP response headers which can't be set or deleted.
+
+## Motivation
+
+Certain HTTP response headers may need to be set, deleted for compliance reasons or other reasons. In OpenShift 3, cluster 
+administrators could use a custom HAProxy configuration template to set or delete headers, but using a custom
+template is not permitted in OpenShift 4.
+Certain HTTP response headers like HSTS (Strict-Transport-Security) need to be denied as we already provide a way to set them
+via route annotation.
+
+### User Stories
+
+1. As a cluster administrator, I need to configure an IngressController to set/delete certain HTTP response headers in order to
+   satisfy compliance requirements at global level to send/not to send them to the client which has called any route from IngressController's domain.
+2. As a user, I need to configure Route to set/delete HTTP response headers to send/not send them to the client which has called the route.
+
+### Goals
+
+1. Enable the cluster administrator to specify HTTP response headers in the IngressController CR spec that should be set or 
+   deleted at the global level, i.e. for all routes.
+2. Enable the user to specify HTTP response headers in the Route spec (when it is not a passthrough Route) for their application
+   that should be set or deleted when forwarding a response to the client.
+3. Define an immutable/hard-coded deny list of headers which are not allowed to be set or deleted.
+   Eg- we should prohibit specifying strict-transport-security using the new API, for the following reasons:
+   HSTS is already configurable through a route annotation.
```

Comment:
IMO, as an API reviewer, I would suggest this isn't a good reason here. We should move away from annotation based APIs where possible

Reply:
Yes. But, removing this now will be a big thing to do. We can target this may be in the next release.

Reply:
It's not just configurable through a route annotation, and that is a better reason why.  There is a whole API for configuring HSTS, at an admin level.